### DISCUSSION TITLE
ROPKinematics for second axis

### DISCUSSION THREAD

**Discussion by savolla** [tesseract-robotics/tesseract#750] [Q&A]
I'm trying to build 8 DOF system (6 for robot and 2 for xy gantry). I think the ROPKinematics class is used for rails which robot can slide on just one axis. But I want to add an additional axis which moves the rail back and forth. Is there a way to implement it? Or do I need to implement another kinematics class with second axis like ROP2?

**Comment by Levi-Armstrong** [comment]
The ROP is what you want to use. The positioner should include the chain for the gantry base up to the frame where the robot mounts and then define the sampling resolution for those two axis. Then define the robot. [Here](https://github.com/tesseract-robotics/tesseract/blob/master/tesseract_support/urdf/online_planning_example_plugins.yaml) is an example using the ROP to define a robot attached to a two axis gantry.

**Reply by savolla** [reply]
Thank you for the quick response! I did what you said. I modified the **workcell_rail_plugins.yaml** like this:
[workcell_rail_plugins.yaml](https://www.toptal.com/developers/hastebin/quwohojina.yaml)
I also wonder where did you generate these .yaml files. Where you defined all the kinematics and plugins. I don't think you wrote them by hand. or is it? :)

Then for the debug purposes I only left 2 points inside **job_path.yaml** like this:
[job_path.yaml](https://www.toptal.com/developers/hastebin/uhogomejoq.yaml)

When I run the example with roslaunch command we get different error messages depending on point position we give inside job_path.yaml. I don't understand the relation between these errors. Here are the error messages:

[ERROR_1](https://www.toptal.com/developers/hastebin/bidosodehe.yaml) *(motion planning server error)*
[ERROR_2](https://www.toptal.com/developers/hastebin/sowigaxuzi.yaml) *(Motion Plan Failed)*

**Comment by Levi-Armstrong** [comment]
> I also wonder where did you generate these .yaml files. Where you defined all the kinematics and plugins. I don't think you wrote them by hand. or is it? :)

Unfortunately I create them by hand by coping from the wiki and making slight modification. In the near future will be able to do this from a GUI.

> Error:   Check Kinematics failed. This means that inverse kinematics solution for a pose do not match forward kinematics solution. Did you change the URDF recently?
         at line 412 in /home/savolla/tesseract_ws/src/tesseract/tesseract_environment/src/environment.cpp

This error indicates that the parameters you have defined for the ROP to not match what is in the URDF. This is checked by creating random joint state get the forward kinematic pose and then running that pose through the IK solver and then its solutions through fwd kinematics and verify that the pose align. Can you post your plugin file?

> Error:   Semantic description is not specified for the same robot as the URDF
         at line 112 in /home/savolla/tesseract_ws/src/tesseract/tesseract_srdf/src/srdf_model.cpp
[ INFO] [1650456050.722750579]: Publishing maintained environment on '/tesseract_workcell_environment/tesseract_published_environment'
Error:   Semantic description is not specified for the same robot as the URDF
         at line 112 in /home/savolla/tesseract_ws/src/tesseract/tesseract_srdf/src/srdf_model.cpp

These mean that the name given to the robot in the URDF and SRDF are not the same.

> Info:    Descartes Processed: 1 of 4 vertices
Info:    Descartes Processed: 2 of 4 vertices
Info:    Descartes Processed: 3 of 4 vertices
Info:    Descartes Processed: 4 of 4 vertices
Debug:   Descartes took 0.9060 seconds to build vertices.
Info:    Descartes Processed: 1 of 3 edges
Info:    Descartes Processed: 2 of 3 edges
Info:    Descartes Processed: 3 of 3 edges
Debug:   Descartes took 0.0012 seconds to build edges.
Warning: Failed vertices:
	1
	2
         at line 62 in /home/savolla/tesseceract_ws/src/descartes_light/descartes_light/core/include/descartes_light/solvers/ladder_graph/impl/ladder_graph_solver.hpp
Warning: Failed edges:
	0
	1
	2

This means that it failed to find an IK solution at vertex 1 and 2 in your toolpath. This is either that they are unreachable or in collision. This may also be do to the issue with the parameters in the IK solver not passing the test. I would run your ik solver through the tesseract unit tests and make sure they are correct. Then we can work on this if an issue after any changes.

**Reply by savolla** [reply]
Thanks for the response! We solved the issue by changing the axis number from 7 to 8 [here](https://github.com/tesseract-robotics/tesseract_ros_workcell/blob/acb32354407ecf79ea9eab37eb5b6edb2d132ca7/twc_motion_planning/src/twc_planning_server_node.cpp#L90-L94). Since in our case the rail now moves back and forth, the axis number remained at 7 inside [twc_planning_server_node.cpp](https://github.com/tesseract-robotics/tesseract_ros_workcell/blob/master/twc_motion_planning/src/twc_planning_server_node.cpp). We also noticed that the main problem was caused by the wrong matrix multiplication because of the old 7 axis configuration. Once we changed it to 8, the application worked flawlessly!

We're now facing a completely different issue with toolpath trajectory. I'm leaving 2 target frames for simplicity inside   [job_path.yaml](https://www.toptal.com/developers/hastebin/ivugojaxev.yaml) file and run the new rail application. Toolpath Frames look like this in Rviz;

<img src="https://user-images.githubusercontent.com/22548773/164661570-d7bdf7fa-f3d2-4ee8-a6a1-77eadad66566.png" alt="drawing" width="500"/>

Here are the gifs representing the weird trajectory;

<img src="https://user-images.githubusercontent.com/22548773/164675444-0ab3d979-1031-4901-8e58-4cb2313488e4.gif" alt="g1" width="700"/>

<img src="https://user-images.githubusercontent.com/22548773/164674159-a9514de8-c718-47b0-aba8-7200cfb7af12.gif" alt="g1" width="700"/>

The generated trajectory is very smooth and nice but  as you can see it jumps on the target frames. I wonder what might cause this behaviour.

**Reply by Levi-Armstrong** [reply]
What process planner are you using?

**Reply by savolla** [reply]
I didn't changed anything in terms of process planner. I used this example workcell as is. The only thing I changed was the urdf/srdf and adding a new axis. I thought Descartes was the process planner or am I wrong? 

EDIT:
Sorry I mixed process and path planners.. I'm using default process planner. I didn't change this.

**Reply by Levi-Armstrong** [reply]
Are you able to post the command language results? You can serialize the results instruction to xml using tools in tesseract_common package serialization.h

**Reply by savolla** [reply]
I finally generated the command launguage results [here](https://www.toptal.com/developers/hastebin/xipugiposi.xml). I inspected the output but couldn't come to any conclusion. I'm using the workcell as is as I mentioned above. The **jump** problem does not occur on the original example.

**Reply by Levi-Armstrong** [reply]
The file provided appears to the input instruction, do you have the results instruction?

**Reply by savolla** [reply]
I'll generate this one as well once I got my hands on my laptop. thank you!

**Reply by savolla** [reply]
Here are the [results](https://gist.github.com/savolla/e9c60fce9f20e685c7488cf7dccb5907)!
I found exact waypoints where the robot **jupms** in between. here is the [last freespace waypoint](https://gist.github.com/savolla/e9c60fce9f20e685c7488cf7dccb5907#file-motion_planning_instructions_results-xml-L7263-L7270) and the [first cartesian waypoint](https://gist.github.com/savolla/e9c60fce9f20e685c7488cf7dccb5907#file-motion_planning_instructions_results-xml-L7405-L7412). Hope it helps to find the problem.

**Reply by Levi-Armstrong** [reply]
I took a look at the results and here is a plot of the trajectory. I would need access to the example to investigate further. Is it available for me to clone down and run?

![image](https://user-images.githubusercontent.com/9803128/166460564-2274da74-e1a5-4dc6-85e4-6b11550afa62.png)
